Make TestEmbeddingNNDeviceTypeCPU::test_EmbeddingBag_per_sample_weights_and_no_offsets_cpu_int32_float32 compatible with TorchDynamo

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120831

Previously, the test case directly accesses the tensor data via tensor.data which is not supported on FakeTensor. So we manually copy the tensor as a workaround. 
Fixes: https://github.com/pytorch/pytorch/issues/119788

tugsbayasgalan committed:
Fix test

[ghstack-poisoned]

janeyx99 commented on test/nn/test_embedding.py:
```diff
@@ -1002,9 +1003,9 @@ def _test_EmbeddingBag_vs_Embedding(self, N, D, B, L, max_norm=None,
         output.backward(grad_output)
         ref_output.backward(grad_output)
-        es_weight_grad = es.weight.grad.data
+        es_weight_grad = copy.deepcopy(es.weight.grad)
```
```suggestion
        es_weight_grad = es.weight.grad.detach()
```

or use clone()?  deepcopy is likely not the right thing to do :p

tugsbayasgalan replied:
ok done

tugsbayasgalan committed:
Update on "Make TestEmbeddingNNDeviceTypeCPU::test_EmbeddingBag_per_sample_weights_and_no_offsets_cpu_int32_float32 compatible with TorchDynamo"


Fixes: https://github.com/pytorch/pytorch/issues/119788


[ghstack-poisoned]

janeyx99 commented on test/nn/test_embedding.py:
```diff
@@ -1002,9 +1002,9 @@ def _test_EmbeddingBag_vs_Embedding(self, N, D, B, L, max_norm=None,
         output.backward(grad_output)
         ref_output.backward(grad_output)
-        es_weight_grad = es.weight.grad.data
+        es_weight_grad = torch.clone(es.weight.grad)
```
Did detach not work lol

tugsbayasgalan replied:
detach had a same error :(

janeyx99 reviewed (commented):
Would trust @mikaylagawarecki's thoughts on whether this tests the same stuff

mikaylagawarecki commented on test/nn/test_embedding.py:
```diff
@@ -1002,9 +1002,9 @@ def _test_EmbeddingBag_vs_Embedding(self, N, D, B, L, max_norm=None,
         output.backward(grad_output)
         ref_output.backward(grad_output)
-        es_weight_grad = es.weight.grad.data
+        es_weight_grad = torch.clone(es.weight.grad)
```
Why did we need the `.data` here initially? 

From running locally, I think just `es.weight.grad` might be ok

mikaylagawarecki replied:
For more context, `.data` calls into `t.variable_data`, impl here 

https://github.com/pytorch/pytorch/blob/aa36821615a4b5cfa748b0fe07fb2f1050af629b/torch/csrc/autograd/variable.cpp#L438-L446

It seems like we just compare `es_weight_grad` with `e.weight.grad` below, so I don't see why this initial `.data` call here would be necessary 🤔

mikaylagawarecki commented:
Separately, I think the other 11 tests when expanding the dropdown in the linked issue might still be failing (probably also due to .data calls in this file) 😅 Would be nice to fix them all!

tugsbayasgalan committed:
Update on "Make TestEmbeddingNNDeviceTypeCPU::test_EmbeddingBag_per_sample_weights_and_no_offsets_cpu_int32_float32 compatible with TorchDynamo"


Previously, the test case directly accesses the tensor data via tensor.data which is not supported on FakeTensor. So we manually copy the tensor as a workaround. 
Fixes: https://github.com/pytorch/pytorch/issues/119788


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Make TestEmbeddingNNDeviceTypeCPU::test_EmbeddingBag_per_sample_weights_and_no_offsets_cpu_int32_float32 compatible with TorchDynamo"


Previously, the test case directly accesses the tensor data via tensor.data which is not supported on FakeTensor. So we manually copy the tensor as a workaround. 
Fixes: https://github.com/pytorch/pytorch/issues/119788


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Make TestEmbeddingNNDeviceTypeCPU::test_EmbeddingBag_per_sample_weights_and_no_offsets_cpu_int32_float32 compatible with TorchDynamo"


Previously, the test case directly accesses the tensor data via tensor.data which is not supported on FakeTensor. So we manually copy the tensor as a workaround. 
Fixes: https://github.com/pytorch/pytorch/issues/119788


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Make TestEmbeddingNNDeviceTypeCPU::test_EmbeddingBag_per_sample_weights_and_no_offsets_cpu_int32_float32 compatible with TorchDynamo"


Previously, the test case directly accesses the tensor data via tensor.data which is not supported on FakeTensor. So we manually copy the tensor as a workaround. 
Fixes: https://github.com/pytorch/pytorch/issues/119788


[ghstack-poisoned]

janeyx99 reviewed (commented):
would it be better to disable assertRaises from within the dynamo code instead vs here?

tugsbayasgalan commented:
> would it be better to disable assertRaises from within the dynamo code instead vs here?

Yepp i think so too. I think we should graph break on assertRaises

janeyx99 commented:
ah i suppose this is also hinged on https://github.com/pytorch/pytorch/pull/119599

janeyx99 commented:
@pytorchbot merge -r

forcing a rebase to see pull get triggered hopefully

tugsbayasgalan committed:
Update

[ghstack-poisoned]
